wallet/src/Pos/Wallet/Web/Methods/History.hs

```diff
@@ -24,42 +24,41 @@ import           Pos.Core                   (getTimestamp, timestampToPosix)
                                              networkChainDifficulty)
-import           Pos.Wallet.Web.ClientTypes (AccountId (..), Addr, CId, CTx (..), CTxId,
-                                             CTxMeta (..), CWAddressMeta (..), Wal, mkCTx)
+import           Pos.Wallet.Web.ClientTypes (AccountId (..), Addr, AddrTxFilter (..), CId,
+                                             CTx (..), CTxId, CTxMeta (..),
+                                             CWAddressMeta (..), Wal, applyAddrTxFilter,
+                                             mkCTx)
 import           Pos.Wallet.Web.Error       (WalletError (..))
 import           Pos.Wallet.Web.Mode        (MonadWalletWebMode)
 import           Pos.Wallet.Web.Pending     (PendingTx (..), ptxPoolInfo)
 import           Pos.Wallet.Web.State       (AddressLookupMode (Ever), addOnlyNewTxMeta,
-                                             getHistoryCache, getPendingTx, getTxMeta,
+                                             getHistoryCachePart, getPendingTx, getTxMeta,
                                              getWalletPendingTxs, setWalletTxMeta)
 import           Pos.Wallet.Web.Util        (decodeCTypeOrFail, getAccountAddrsOrThrow,
                                              getWalletAccountIds, getWalletAddrMetas,
                                              getWalletAddrs, getWalletThTime)
 
 
-getFullWalletHistory :: MonadWalletWebMode m => CId Wal -> m ([CTx], Word)
-getFullWalletHistory cWalId = do
+-- | Gather history from all sources: db, mempool txs, pending txs.
+-- Note that @minNum@ is just a prompt about minimal number of
+-- transactions to return.
+getFullWalletHistory
+    :: MonadWalletWebMode m
+    => CId Wal -> AddrTxFilter -> Int -> m ([CTx], Word)
+getFullWalletHistory cWalId filtering minNum = do
     addrs <- mapM decodeCTypeOrFail =<< getWalletAddrs Ever cWalId
 
-    unfilteredLocalHistory <- getLocalHistory addrs
+    unfilteredLocalHistory <-
+        applyAddrTxFilter filtering . DL.toList <$> getLocalHistory addrs
 
-    blockHistory <- getHistoryCache cWalId >>= \case
-        Just hist -> pure $ DL.fromList hist
-        Nothing -> do
-            logWarning $
-                sformat ("getFullWalletHistory: history cache is empty for wallet #"%build)
-                cWalId
-            pure mempty
+    blockHistory <- getHistoryCachePart filtering minNum cWalId
 
-    let localHistory =
-            DL.fromList $ filterLocalTh
-                (DL.toList blockHistory)
-                (DL.toList unfilteredLocalHistory)
+    let localHistory = filterLocalTh blockHistory unfilteredLocalHistory
 
     logTxHistory "Block" blockHistory
     logTxHistory "Mempool" localHistory
 
-    fullHistory <- addRecentPtxHistory cWalId $ DL.toList $ localHistory <> blockHistory
+    fullHistory <- addRecentPtxHistory cWalId $ localHistory <> blockHistory
     cHistory <- forM fullHistory $ addHistoryTx cWalId
```

Comment:
I don't get why do we add it here? what does `addHistoryTx` do? is it really add some txs to db?

Reply:
anyway I don't line that functions with prefix `get` does some modification...

Reply:
> anyway I don't line that functions with prefix get does some modification...

Yes!

Reply:
I assume this line affects performance b/c we write to db on each transaction

Reply:
I tried with 10k transactions, all before `addHistory` calls passed instantly, while these calls took 24 secs